keps/sig-autoscaling/0000-kep-template.md

```diff
@@ -0,0 +1,273 @@
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+
+### Goals
+
+* cluster autoscaler is capable of scaling the cluster through machine API
+
+### Non-Goals
+
+* cluster autoscaler is able to autoprovision new node groups through machine API
+* cluster autoscaler is able to estimate node resource requirements through machine API
+* cluster autoscaler is able to use at least one pricing model on top of the machine API
+
+## Proposal
+
+Generalize the concept of node group autoscaling on the level of cloud provider
+by integrating with machine API layer of [sigs.k8s.io/cluster-api](https://github.com/kubernetes-sigs/cluster-api) project to
+build cloud provider-free implementation of cluster autoscaling mechanism
+([#1050](https://github.com/kubernetes/autoscaler/issues/1050)).
+
+Also suggest how to:
+* perform node autoprovision on the level of the machine API ([node autoprovisioning
+  upstream proposal](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/proposals/node_autoprovisioning.md)).
+* perform estimation of cluster autoscaling on the level of the machine API.
+* build various pricing models on top of the machine API.
+
+Cluster autoscaling builds on top of concept of node groups which effectively
+translates to cloud provider specific autoscaling groups, resp. autoscaling sets.
+In case of the machine API, the node group translates into machine set (living
+as native Kubernetes object).
+
+### Node group (scaling group)
+
+The autoscaler currently supports `AWS`, `Azure`, `GCE`, `GKE` providers.
+With `kubemark` as other cloud provider implementation that allows to create
+hollow kubernetes cluster with huge number of nodes over a small subset of real
+machines (for scaling testing purposes).
+Each of the currently supported cloud providers has its own concept of scaling groups:
+- AWS has autoscaling groups
+- Azure has autoscaling sets
+- GCE/GKE have autoscaling groups
+
+Autoscaling group allows to flexibly and natively increase/decrease number of instances
+within cloud provider based on actual cluster workload.
+
+TODO(jchaloup): describe advantages of auto scaling groups, why they are beneficial, etc.
+
+### Machine API
+
+Machine API (as part of the [cluster-api project](https://github.com/kubernetes-sigs/cluster-api/))
+allows to declaratively specify a machine without saying how to provision underlying instance.
+Machines can be grouped into a machine set that allows to specify a machine template
+and by setting a number replicas create required number of machines (followed
+by provisioning of underlying instances).
+With some mama's marmalade, one can represent an autoscaling group with a machine set.
+Though, the current implementation of the machineset controller needs to be improved
+to support autoscaling use cases. E.g. define delete strategy to pick appropriate machine
```

Comment:
It's crucial to Cluster Autoscaler logic that it is able to name a specific machine to be deleted in case of scale-down. Delete strategy on the side of machine set alone won't work for this use-case.

Reply:
I think we should have a write-up on why this is the case, since it doesn't seem to be immediately obvious.

More specifically, it seems at first glance that the machineset controller would want similar criteria for scale-down to  the cluster autoscaler; a scale-down that doesn't take into account disruption budget, node load, node readiness, particular pod types, etc doesn't seem like it would be particularly useful.

The difference that jumps to mind initially is the scheduling simulation -- is the issue that the optimal node to delete in terms of load+disruption-budget+pod-types+etc might have pods that don't fit elsewhere?

Reply:
I'm not sure if machine set should have all this logic. If the user resizes it some node must be removed, even if doing so violates PDB, kills a pod that won't be re-created, etc. There may be no right decision and any heuristic would be arbitrary. 

Given the ambiguity I think it's more correct to put the responsibility to choose node on whoever makes the decision to remove it (CA, user, other controller). If no explicit decision is made machine set should use some very simple and predictable strategy (remove newest node, or oldest node, or something like that). Ideally the API should allow to reduce target size of machine deployment and remove specific machine. Some new subresource maybe?

BTW it feels like stateful set would have the same problem. Without detailed knowledge about workload it's not possible to make an informed decision on which replica to remove, so a user really should have an option to choose. It feels like Cluster API should do the same thing stateful set does (or the other way round, if stateful set haven't solved it yet).

Reply:
Also that's just how CA works. It picks a node to delete, not a node group to resize. Unless we guarantee the internal logic is 100% the same between CA and machine controller they will conflict.

Reply:
> Given the ambiguity I think it's more correct to put the responsibility to choose node on whoever makes the decision to remove it (CA, user, other controller)

That may be an acceptable answer, but I think we should *explicitly* say that in the proposal, and enumerate why a bit.  I've seen a lot of "it has to work this way because the cluster autoscaler *currently* works this way" written down, and I think we need to write down more justification beyond that.

> It feels like Cluster API should do the same thing stateful set does (or the other way round, if stateful set haven't solved it yet).

Yeah, I wonder if the workloads folks have solved that yet @kubernetes/sig-apps-misc 

> Also that's just how CA works

If we're designing for future plans, maybe we should re-evaluate that.  Maybe the CA should be closer to the way the HPA works? (i.e. "have we re-evaluated our choices in light of the cluster API being a thing")

Reply:
> Maybe the CA should be closer to the way the HPA works? 

Various groups of VMs (ASGs on AWS, MIGs on GCP etc.) already work similarly to HPA (i.e., the number of replicas depends on a metric's value). So at least in those environment, users can already have it without running a separate component. The problem is that autoscaling Kubernetes nodes in this way doesn't work well in most cases. In fact, it causes so much trouble we've had to put explicit warnings in UI against enabling MIG autoscaler for VMs running GKE nodes.

It all boils down to "what does it mean that we need a bigger/smaller cluster". If a metric we use as input is above target, but there are no pending pods, adding a new machine will do nothing except unnecessarily increase the cost of running the cluster, because nothing will be scheduled there (although it may indeed bring some metrics down, e.g. average utilization.) This is even worse for scale-down - not all nodes run the same workloads, and removing nodes without checking the pods running there can be moved elsewhere may result in some of those pods no longer running (possibly taking down entire applications.)

It can be argued we just haven't figured out the right metrics for it yet, but even if we could use something like "number of pending pods" as input to scale-up, I've yet to see anything that would work for scale-down. And if we find one, perhaps it can be used with already existing VM autoscalers?